File: backend/app/months.py
"""Month-key utilities — a Python port of `src/utils/months.ts`.

A "month key" is the string `YYYY-MM`. The ledger range defined here must stay
in sync with the front end's `LEDGER_START` / `LEDGER_END`; both sides derive
their month pickers and validation from it.
"""

import re
from calendar import monthrange
from datetime import date

LEDGER_START = "2026-07"
LEDGER_END = "2027-12"

_MONTH_KEY_PATTERN = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")


def is_valid_month_key(key: str) -> bool:
    return bool(_MONTH_KEY_PATTERN.match(key))


def parse_month_key(key: str) -> tuple[int, int]:
    year, month = key.split("-")
    return int(year), int(month)


def to_month_key(year: int, month: int) -> str:
    return f"{year}-{month:02d}"


def _month_index(key: str) -> int:
    year, month = parse_month_key(key)
    return year * 12 + (month - 1)


def _from_month_index(index: int) -> str:
    return to_month_key(index // 12, (index % 12) + 1)


def compare_month_keys(a: str, b: str) -> int:
    return _month_index(a) - _month_index(b)


def is_within_ledger_range(key: str) -> bool:
    return (
        is_valid_month_key(key)
        and compare_month_keys(key, LEDGER_START) >= 0
        and compare_month_keys(key, LEDGER_END) <= 0
    )


def shift_month_key(key: str, delta: int) -> str | None:
    nxt = _from_month_index(_month_index(key) + delta)
    return nxt if is_within_ledger_range(nxt) else None


def month_key_range(start: str = LEDGER_START, end: str = LEDGER_END) -> list[str]:
    return [_from_month_index(i) for i in range(_month_index(start), _month_index(end) + 1)]


def month_keys_in_year(year: int) -> list[str]:
    return [key for key in month_key_range() if parse_month_key(key)[0] == year]


def month_key_from_date(value: date | str) -> str:
    if isinstance(value, date):
        return value.strftime("%Y-%m")
    return value[:7]


def days_in_month(key: str) -> int:
    year, month = parse_month_key(key)
    return monthrange(year, month)[1]


def iso_date_in(month_key: str, day: int) -> str:
    clamped = min(max(day, 1), days_in_month(month_key))
    return f"{month_key}-{clamped:02d}"


def current_snapshot_month(today: date | None = None) -> str:
    """The month whose net-worth point is computed live rather than stored.

    Clamped into the ledger range so a Pi with a wrong clock, or use before the
    ledger opens, still resolves to a real month.
    """
    today = today or date.today()
    key = to_month_key(today.year, today.month)
    if compare_month_keys(key, LEDGER_START) < 0:
        return LEDGER_START
    if compare_month_keys(key, LEDGER_END) > 0:
        return LEDGER_END
    return key

File: frontend/src/utils/months.ts
/**
 * Month-key utilities.
 *
 * A "month key" is the string `YYYY-MM` (e.g. `2026-07`). It is the identifier
 * used in routes (`/months/2026-07`) and by the API layer. The backend derives
 * the same key from a transaction's date.
 */

/** Ledger data starts here. Nothing before this month exists, ever. */
export const LEDGER_START = '2026-07'

/**
 * Last month offered by the month picker. Extend this to add more months —
 * everything (picker, yearly table year list, month navigation) derives from it.
 */
export const LEDGER_END = '2027-12'

export const MONTH_NAMES = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
] as const

export const MONTH_NAMES_SHORT = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
] as const

const MONTH_KEY_PATTERN = /^\d{4}-(0[1-9]|1[0-2])$/

export function isValidMonthKey(key: string): boolean {
  return MONTH_KEY_PATTERN.test(key)
}

/** Splits `2026-07` into `{ year: 2026, month: 7 }` (month is 1-indexed). */
export function parseMonthKey(key: string): { year: number; month: number } {
  const [year, month] = key.split('-')
  return { year: Number(year), month: Number(month) }
}

export function toMonthKey(year: number, month: number): string {
  return `${year}-${String(month).padStart(2, '0')}`
}

/** Converts a month key to a sortable integer, e.g. `2026-07` -> 24319. */
function monthIndex(key: string): number {
  const { year, month } = parseMonthKey(key)
  return year * 12 + (month - 1)
}

export function compareMonthKeys(a: string, b: string): number {
  return monthIndex(a) - monthIndex(b)
}

export function isWithinLedgerRange(key: string): boolean {
  return (
    isValidMonthKey(key) &&
    compareMonthKeys(key, LEDGER_START) >= 0 &&
    compareMonthKeys(key, LEDGER_END) <= 0
  )
}

/** Adds `delta` months to a key. Returns `null` if the result leaves the ledger range. */
export function shiftMonthKey(key: string, delta: number): string | null {
  const index = monthIndex(key) + delta
  const next = toMonthKey(Math.floor(index / 12), (index % 12) + 1)
  return isWithinLedgerRange(next) ? next : null
}

/** All month keys from `from` to `to` inclusive. */
export function monthKeyRange(from = LEDGER_START, to = LEDGER_END): string[] {
  const keys: string[] = []
  for (let i = monthIndex(from); i <= monthIndex(to); i++) {
    keys.push(toMonthKey(Math.floor(i / 12), (i % 12) + 1))
  }
  return keys
}

/** Every year that has at least one month in the ledger range. */
export function ledgerYears(): number[] {
  const years = new Set(monthKeyRange().map((key) => parseMonthKey(key).year))
  return [...years].sort((a, b) => a - b)
}

/** Month keys within the ledger range that fall in `year`. */
export function monthKeysInYear(year: number): string[] {
  return monthKeyRange().filter((key) => parseMonthKey(key).year === year)
}

/** `2026-07` -> `July 2026` */
export function formatMonthLong(key: string): string {
  const { year, month } = parseMonthKey(key)
  return `${MONTH_NAMES[month - 1]} ${year}`
}

/** `2026-07` -> `Jul 2026` */
export function formatMonthShort(key: string): string {
  const { year, month } = parseMonthKey(key)
  return `${MONTH_NAMES_SHORT[month - 1]} ${year}`
}

/** The month key an ISO date (`2026-07-14`) belongs to. */
export function monthKeyFromDate(isoDate: string): string {
  return isoDate.slice(0, 7)
}

/** Number of days in the given month key. */
export function daysInMonth(key: string): number {
  const { year, month } = parseMonthKey(key)
  return new Date(year, month, 0).getDate()
}

/** `2026-07-14` -> `Jul 14` */
export function formatDayShort(isoDate: string): string {
  const [, month, day] = isoDate.split('-')
  return `${MONTH_NAMES_SHORT[Number(month) - 1]} ${Number(day)}`
}

/** Builds an ISO date string inside a month, clamped to that month's length. */
export function isoDateIn(monthKey: string, day: number): string {
  const clamped = Math.min(Math.max(day, 1), daysInMonth(monthKey))
  return `${monthKey}-${String(clamped).padStart(2, '0')}`
}
